src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/Utilities/AsyncEventMap.cs

```diff
@@ -0,0 +1,73 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Threading.Tasks;
+
+namespace Roslyn.Utilities;
+
+internal sealed class AsyncEventMap
+{
+    private readonly NonReentrantLock _guard = new();
+    private readonly Dictionary<string, object> _eventNameToHandlerSet = [];
+
+    public void AddAsyncEventHandler<TEventArgs>(string eventName, Func<TEventArgs, Task> asyncHandler)
+        where TEventArgs : EventArgs
+    {
+        using (_guard.DisposableWait())
+        {
+            _eventNameToHandlerSet[eventName] = _eventNameToHandlerSet.TryGetValue(eventName, out var handlers)
+                ? ((AsyncEventHandlerSet<TEventArgs>)handlers).WithHandler(asyncHandler)
+                : new AsyncEventHandlerSet<TEventArgs>([asyncHandler]);
+        }
+    }
+
+    public void RemoveAsyncEventHandler<TEventArgs>(string eventName, Func<TEventArgs, Task> asyncHandler)
+        where TEventArgs : EventArgs
+    {
+        using (_guard.DisposableWait())
+        {
+            if (_eventNameToHandlerSet.TryGetValue(eventName, out var handlers))
+                _eventNameToHandlerSet[eventName] = ((AsyncEventHandlerSet<TEventArgs>)handlers).WithHandlerRemoved(asyncHandler);
```

Comment:
can we do .AddHandler/.RemoveHandler?  or .WithHandlerAdded/.WithHandlerRemoved.  The name inconsistency bothers  me :)